addon/vector.js

```diff
@@ -74,7 +74,39 @@ class CorpusCollector extends PageVisitor {
         }
         if (vector !== undefined) {
             this._vectors.push(vector);
-            this.setCurrentStatus({message: 'vectorized', isFinal: true});
+
+            // Check if any of the rules didn't run or returned null.
+            // This presents as an undefined value in a feature vector.
+            const nullFeatures = await this.nullFeatures(vector.nodes);
+            if (nullFeatures) {
+                console.log(nullFeatures);
+                this.setCurrentStatus({
+                    message: 'failed: rule(s) ' + nullFeatures + ' returned null values',
+                    isFinal: true
+                });
+            } else {
+                this.setCurrentStatus({message: 'vectorized', isFinal: true});
+            }
+        }
+    }
+
+    async nullFeatures(nodes) {
+        for (const node of nodes) {
+            if (node.features.some(featureValue => featureValue === undefined)) {
+                // TODO: Should we get the feature names elsewhere so we don't have to make this async call multiple times?
```

Comment:
Good observation.

Certainly the first time it is called, we could store reusable properties on the class instance and check those properties before each message sending opportunity. E.g. `this.featureNames`, `this.viewportWidth`, `this.viewportHeight`, ...).

At a minimum, in `nullFeatures` itself, you could move the messaging part and `featureNames` assignment outside of the `for` loop:

```js
    async nullFeatures(nodes) {
        const trainee = await browser.runtime.sendMessage(
            'fathomtrainees@mozilla.com',
            {type: 'trainee', traineeId: this.otherOptions.traineeId},
        );
        const featureNames = Array.from(trainee.coeffs.keys());
        for (const node of nodes) {
            if (node.features.some(featureValue => featureValue === undefined)) {
                return node.features.reduce((nullFeatures, featureValue, index) => {
                    if (featureValue === undefined) {
                        nullFeatures.push(featureNames[index]);
                    }
                    return nullFeatures;
                }, []);
            }
        }
    }
```

We could also make the code DRYer by having that message sending block to fetch `trainee` in only one place -- I see two other calls at least elsewhere in `vector.js` for getting the `trainee` object.

Reply:
This is a really good point. I think I have a nice changeset lined up to address this. It has an added benefit of making several methods no longer async.